# Exploring Hacker News Posts

In this project we are going to work with a data set of submissions to popular technology site [Hacker News](HN_posts_year_to_Sep_26_2016.csv).

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set [here](https://www.kaggle.com/hacker-news/hacker-news-posts), but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. 

In this mission, we'll compare 'Ask HN' and 'Show HN' those two types of posts to discover how their numbers of comment related with average numbers of comment as well as the posts' creating date. 

<br/>
<br/>
Below are descriptions of the columns:

**id**: The unique identifier from Hacker News for the post

**title**: The title of the post

**url**: The URL that the posts links to, if it the post has a URL

**num_points**: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes

**num_comments**: The number of comments that were made on the post

**author**: The username of the person who submitted the post


**created_at**: The date and time at which the post was submitted

In [60]:
from csv import reader
opened_file = open('HN_posts_year_to_Sep_26_2016.csv/HN_posts_year_to_Sep_26_2016.csv', encoding='utf8')
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0:1]
hn = hn[1:]

In [61]:
def explore_data(data_set,start,end):
    data_set_slice = data_set[start:end]
    for row in data_set_slice:
        print(row)
        print('\n')
    
    print('The Numbers of row: {}'.format(len(data_set)))
    print('The numbrts of column: {}'.format(len(data_set[0])))

In [62]:
print(headers)
print('\n')
explore_data(hn,0,5)

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]


['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26']


['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24']


['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19']


['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16']


['12578979', 'How the Data Vault Enables the Next-Gen Data Warehouse and Data Lake', 'https://www.talend.com/blog/2016/05/12/talend-and-Â\x93the-data-vaultÂ\x94', '1', '0', 'markgainor1', '9/26/2016 3:14']


The Numb

In [63]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [64]:
print(len(ask_posts))
print('\n')
print(len(show_posts))
print('\n')
print(len(other_posts))

9139


10158


273822


In [65]:
print(*ask_posts[0:5],sep='\n')

['12578908', 'Ask HN: What TLD do you use for local development?', '', '4', '7', 'Sevrene', '9/26/2016 2:53']
['12578522', 'Ask HN: How do you pass on your work when you die?', '', '6', '3', 'PascLeRasc', '9/26/2016 1:17']
['12577908', 'Ask HN: How a DNS problem can be limited to a geographic region?', '', '1', '0', 'kuon', '9/25/2016 22:57']
['12577870', 'Ask HN: Why join a fund when you can be an angel?', '', '1', '3', 'anthony_james', '9/25/2016 22:48']
['12577647', 'Ask HN: Someone uses stock trading as passive income?', '', '5', '2', '00taffe', '9/25/2016 21:50']


In [66]:
print(*show_posts[0:5],sep='\n')

['12578335', 'Show HN: Finding puns computationally', 'http://puns.samueltaylor.org/', '2', '0', 'saamm', '9/26/2016 0:36']
['12578182', 'Show HN: A simple library for complicated animations', 'https://christinecha.github.io/choreographer-js/', '1', '0', 'christinecha', '9/26/2016 0:01']
['12578098', 'Show HN: WebGL visualization of DNA sequences', 'http://grondilu.github.io/dna.html', '1', '0', 'grondilu', '9/25/2016 23:44']
['12577991', 'Show HN: Pomodoro-centric, heirarchical project management with ES6 modules', 'https://github.com/jakebian/zeal', '2', '0', 'dbranes', '9/25/2016 23:17']
['12577142', 'Show HN: Jumble  Essays on the go #PaulInYourPocket', 'https://itunes.apple.com/us/app/jumble-find-startup-essay/id1150939197?ls=1&mt=8', '1', '1', 'ryderj', '9/25/2016 20:06']


Function to work out the average number of costs for data set

In [67]:
def average_num_post(data_set):
    total_comments = 0
    for row in data_set:
        num_comments = int(row[4])
        total_comments  += num_comments
    avg_comments = total_comments / len(data_set)
    return 'Average number of comments is: {:.2f}'.format(avg_comments)
    

Compute the number of comments for 'ask_post'

In [68]:
average_num_post(ask_posts)

'Average number of comments is: 10.39'

Compute the average number of comments for 'show_post'

In [69]:
average_num_post(show_posts)

'Average number of comments is: 4.89'

Compute the average number of comments for 'other_posts'

In [70]:
average_num_post(other_posts)

'Average number of comments is: 6.46'


After calculate the avearge number of comments for 'ask_post', 'show_post' and the whole data set 'hn', we can see the average number of comments for 'hn' (the whole data set) is *6.46* which is between 'ask_post'(*10.39* - the highest) and  'show_post'(*4.89*)

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

Calculate the amount of "**ask posts**" created in each hour of the day, along with the number of comments received.

In [115]:
import datetime as dt

In [72]:
result_list = []

for row in ask_posts:
    sub_list = []
    
    create_time = row[6]
    sub_list.append(create_time)
    
    num_comments = int(row[4])
    sub_list.append(num_comments)
    
    result_list.append(sub_list)

In [82]:
result_list[0:10]

[['9/26/2016 2:53', 7],
 ['9/26/2016 1:17', 3],
 ['9/25/2016 22:57', 0],
 ['9/25/2016 22:48', 3],
 ['9/25/2016 21:50', 2],
 ['9/25/2016 19:30', 1],
 ['9/25/2016 19:22', 22],
 ['9/25/2016 17:55', 3],
 ['9/25/2016 15:48', 0],
 ['9/25/2016 15:35', 13]]

In [80]:
len(result_list)

9139

In [81]:
result_list[9130:]

[['9/6/2015 15:09', 0],
 ['9/6/2015 14:53', 6],
 ['9/6/2015 13:01', 9],
 ['9/6/2015 12:17', 0],
 ['9/6/2015 11:27', 0],
 ['9/6/2015 10:52', 1],
 ['9/6/2015 10:46', 4],
 ['9/6/2015 9:36', 20],
 ['9/6/2015 6:02', 20]]

In [160]:

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    the_date = row[0]
    num_comments = row[1]
    
    the_date_object = dt.datetime.strptime(the_date, '%m/%d/%Y %H:%M')
    the_hour_object = the_date_object.strftime('%H')

    if the_hour_object not in counts_by_hour:
        counts_by_hour[the_hour_object] = 1
        comments_by_hour[the_hour_object] = num_comments
    else:
        counts_by_hour[the_hour_object] += 1
        comments_by_hour[the_hour_object] += num_comments

"""    
    if the_hour_object not in counts_by_hour:
        counts_by_hour[the_hour_object] = 1
    else:
        counts_by_hour[the_hour_object] += 1
    
    if the_hour_object not in comments_by_hour:
        comments_by_hour[the_hour_object] = num_comments
    else:
        comments_by_hour[the_hour_object] += num_comments
"""        
        
print(counts_by_hour)
print('\n')
print(comments_by_hour)



{'02': 269, '01': 282, '22': 383, '21': 518, '19': 552, '17': 587, '15': 646, '14': 513, '13': 444, '11': 312, '10': 282, '09': 222, '07': 226, '03': 271, '23': 343, '20': 510, '16': 579, '08': 257, '00': 301, '18': 614, '12': 342, '04': 243, '06': 234, '05': 209}


{'02': 2996, '01': 2089, '22': 3372, '21': 4500, '19': 3954, '17': 5547, '15': 18525, '14': 4972, '13': 7245, '11': 2797, '10': 3013, '09': 1477, '07': 1585, '03': 2154, '23': 2297, '20': 4462, '16': 4466, '08': 2362, '00': 2277, '18': 4877, '12': 4234, '04': 2360, '06': 1587, '05': 1838}


In [75]:
print(type(counts_by_hour['02']))

<class 'int'>


Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

In [76]:
avg_by_hour = []
for count in counts_by_hour:
    for comment in comments_by_hour:
        average = 0
        if count == comment:
            average =  comments_by_hour[comment] / counts_by_hour[count]
            sub_list = [count, average]
            avg_by_hour.append(sub_list)

        

In [83]:
avg_by_hour

[['02', 11.137546468401487],
 ['01', 7.407801418439717],
 ['22', 8.804177545691905],
 ['21', 8.687258687258687],
 ['19', 7.163043478260869],
 ['17', 9.449744463373083],
 ['15', 28.676470588235293],
 ['14', 9.692007797270955],
 ['13', 16.31756756756757],
 ['11', 8.96474358974359],
 ['10', 10.684397163120567],
 ['09', 6.653153153153153],
 ['07', 7.013274336283186],
 ['03', 7.948339483394834],
 ['23', 6.696793002915452],
 ['20', 8.749019607843136],
 ['16', 7.713298791018998],
 ['08', 9.190661478599221],
 ['00', 7.5647840531561465],
 ['18', 7.94299674267101],
 ['12', 12.380116959064328],
 ['04', 9.7119341563786],
 ['06', 6.782051282051282],
 ['05', 8.794258373205741]]

we calculated the average number of comments for posts created during each hour of the day, and stored the results in a list of lists named "avg_by_hour"

Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [87]:
swap_avg_by_hour = []
for x in avg_by_hour:
    swap_avg_by_hour.append((x[1],x[0]))

swap_avg_by_hour

[(11.137546468401487, '02'),
 (7.407801418439717, '01'),
 (8.804177545691905, '22'),
 (8.687258687258687, '21'),
 (7.163043478260869, '19'),
 (9.449744463373083, '17'),
 (28.676470588235293, '15'),
 (9.692007797270955, '14'),
 (16.31756756756757, '13'),
 (8.96474358974359, '11'),
 (10.684397163120567, '10'),
 (6.653153153153153, '09'),
 (7.013274336283186, '07'),
 (7.948339483394834, '03'),
 (6.696793002915452, '23'),
 (8.749019607843136, '20'),
 (7.713298791018998, '16'),
 (9.190661478599221, '08'),
 (7.5647840531561465, '00'),
 (7.94299674267101, '18'),
 (12.380116959064328, '12'),
 (9.7119341563786, '04'),
 (6.782051282051282, '06'),
 (8.794258373205741, '05')]

In [90]:
swap_avg_by_hour.sort(reverse = True)

In [92]:
sorted_swap = swap_avg_by_hour

In [93]:
sorted_swap

[(28.676470588235293, '15'),
 (16.31756756756757, '13'),
 (12.380116959064328, '12'),
 (11.137546468401487, '02'),
 (10.684397163120567, '10'),
 (9.7119341563786, '04'),
 (9.692007797270955, '14'),
 (9.449744463373083, '17'),
 (9.190661478599221, '08'),
 (8.96474358974359, '11'),
 (8.804177545691905, '22'),
 (8.794258373205741, '05'),
 (8.749019607843136, '20'),
 (8.687258687258687, '21'),
 (7.948339483394834, '03'),
 (7.94299674267101, '18'),
 (7.713298791018998, '16'),
 (7.5647840531561465, '00'),
 (7.407801418439717, '01'),
 (7.163043478260869, '19'),
 (7.013274336283186, '07'),
 (6.782051282051282, '06'),
 (6.696793002915452, '23'),
 (6.653153153153153, '09')]

In [125]:
print('Top 5 Hours for Ask Posts Comments')

for x in sorted_swap[0:5]:
    hour = x[1]
    hour = dt.datetime.strptime(hour, '%H')
    hour = hour.strftime('%H:%M')
    
    print('{}: {:.2f} average comments per post'.format(hour,x[0]))
    
    

Top 5 Hours for Ask Posts Comments
15:00: 28.68 average comments per post
13:00: 16.32 average comments per post
12:00: 12.38 average comments per post
02:00: 11.14 average comments per post
10:00: 10.68 average comments per post


We have already calculated the amount of "**ask posts**" created in each hour of the day, along with the number of comments received.

But, we are still going to continue find the information for "**show posts**" and "**other posts**", so there are duplicate code we need to use to finish our task, we can define a function for this proceess to make it easer for furture use.

In [180]:
#Funtion to calculate the number of post for top 5 rating hours 

#calculate the result_list
def the_result(post_group):
    result_list = []
    for row in post_group:
        create_time = row[6]
        num_comments = int(row[4])
        result_list.append((create_time,num_comments))
    return counts_comments(result_list)

#calculate dictionary for counts_by_hour and comments_by_hour
def counts_comments(result_list):
    counts_by_hour = {}
    comments_by_hour = {}
    
    for row in result_list:
        the_date = row[0]
        num_comments = row[1]
    
        the_date_object = dt.datetime.strptime(the_date, '%m/%d/%Y %H:%M')
        the_hour_object = the_date_object.strftime('%H')

        if the_hour_object not in counts_by_hour:
                counts_by_hour[the_hour_object] = 1
                comments_by_hour[the_hour_object] = num_comments
        else:
                counts_by_hour[the_hour_object] += 1
                comments_by_hour[the_hour_object] += num_comments

        
    return avg_comments(counts_by_hour, comments_by_hour)

#calculate the average post for each hour and print out the top 5 rating hour post number
def avg_comments(counts_by_hour, comments_by_hour):
    avg_by_hour = []
    for count in counts_by_hour:
        for comment in comments_by_hour:
            average = 0
            if count == comment:
                average = comments_by_hour[comment] / counts_by_hour[count]
                avg_by_hour.append((count,average))
    
    swap_avg_by_hour = []
    for x in avg_by_hour:
        swap_avg_by_hour.append((x[1],x[0]))
    
    swap_avg_by_hour.sort(reverse = True)
    sorted_swap = swap_avg_by_hour
                                
    print('Top 5 Hours for Ask Posts Comments')
    for x in sorted_swap[0:5]:
        hour = x[1]
        hour = dt.datetime.strptime(hour, '%H')
        hour = hour.strftime('%H:%M')   
        print('{}: {:.2f} average comments per post'.format(hour,x[0]))
    


Calculate for 'ask_posts'

In [182]:
#after run the 'the_result function, return to get average number of comments by each hour 
the_result(ask_posts) 

Top 5 Hours for Ask Posts Comments
15:00: 28.68 average comments per post
13:00: 16.32 average comments per post
12:00: 12.38 average comments per post
02:00: 11.14 average comments per post
10:00: 10.68 average comments per post


Calculate for 'show_posts'

In [185]:
#after run the 'the_result function, return to get average number of comments by each hour 
the_result(show_posts) 

Top 5 Hours for Ask Posts Comments
12:00: 6.99 average comments per post
07:00: 6.68 average comments per post
11:00: 6.00 average comments per post
08:00: 5.60 average comments per post
14:00: 5.52 average comments per post


Calculate for 'other_posts'

In [186]:
#after run the 'the_result function, return to get average number of comments by each hour 
the_result(other_posts) 

Top 5 Hours for Ask Posts Comments
12:00: 7.59 average comments per post
11:00: 7.37 average comments per post
02:00: 7.18 average comments per post
13:00: 7.15 average comments per post
05:00: 6.79 average comments per post
